Block patterns: reregister remote core and Gutenberg patterns using site locale

ramonjd opened the pull request:
## Changes proposed in this Pull Request

Core [WordPress](https://github.com/WordPress/WordPress/blob/master/wp-includes/block-patterns.php) and [Gutenberg](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-patterns.php#L245
) register patterns behind a `get_theme_support( 'core-block-patterns' )` check.


This PR:

1. deactivates `'core-block-patterns'` theme support to prevent Gutenberg registering its patterns and fudging around with core patterns. This method replaces the string check of pattern names, thereby simplifying the additions we made in https://github.com/Automattic/wp-calypso/pull/52656, and should be more reliable as a result.
2. Loads WordPress core patterns, Gutenberg and remote core patterns manually so we can switch to the site locale first


## Testing instructions

Apply this ETK patch to your sandbox

`install-plugin.sh etk update/remove-core-block-patterns-theme-support`

and open up block editor in a sandboxed site.

Check that we're loading WPCOM patterns and Gutenberg patterns from https://github.com/WordPress/gutenberg/blob/trunk/lib/block-patterns.php.

In the Site Editor, check that we're loading remote core patterns. See equivalent: https://github.com/WordPress/gutenberg/blob/trunk/lib/block-patterns.php#L248

Switch your site language and make sure that you see evidence that the patterns/pattern categories are at least trying to be translated :)

<img width="316" alt="Screen Shot 2021-06-21 at 12 53 05 pm" src="https://user-images.githubusercontent.com/6458278/122700974-a37d8280-d28f-11eb-9701-359207784c05.png">



```
cd apps/editing-toolkit
yarn run test:php --testsuite block-patterns

```

ramonjd committed:
Initial commit. Deactivate 'core-block-patterns' theme support to prevent Gutenberg registering its patterns and fudging around with core patterns.

This method replaces the string check of pattern names, so it's cleaner and might be more reliable.

andrewserong commented:
Nice idea @ramonjd! I'm not sure if I'm missing a step while testing, but it looks like the Query patterns from Gutenberg are still showing on my test site:

![image](https://user-images.githubusercontent.com/14988353/120123295-5b6bc280-c1f1-11eb-822b-17789ac7becc.png)

I'm wondering if the `after_setup_theme` hook is too late in the process for disabling the theme support? It looks like Gutenberg registers its patterns on [init](https://github.com/WordPress/gutenberg/blob/40f058bad84c4c4bb11885bc4dfe37382c027ba6/lib/block-patterns.php#L237)?

ramonjd commented:
> I'm wondering if the after_setup_theme hook is too late in the process for disabling the theme support? It looks like Gutenberg registers its patterns on init?

Thanks for testing! Hmmm, that sounds legit. 

I didn't see anything but only tested once on Friday afternoon and may been "lucky" in that race :) 

I'll take another look.

ramonjd committed:
Spelling.
Moving the remove_theme_support into load_wpcom_block_patterns_modifications because it's already there and works

ramonjd commented:
Okay! So testing on edge (10.7.1) works because Gutenberg 10.5.4 (prod) performs no checks before loading patterns.

https://github.com/WordPress/gutenberg/blob/v10.5.4/lib/block-patterns.php

Thankfully, later versions do:

https://github.com/WordPress/gutenberg/blob/release/10.7/lib/block-patterns.php#L244

So we can put this on hold until WordPress moves over to edge.

andrewserong commented:
Ah, good find! Thanks for confirming 🙂

andrewserong commented on apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php:
```diff
@@ -282,6 +282,12 @@ function load_wpcom_block_patterns_modifications() {
 	// 	require_once __DIR__ . '/block-patterns/class-block-patterns-modifications.php';
 	// }
 	// phpcs:enable
+	/*
```
For when you get back to this PR, PHPCS is complaining that there isn't an empty line above this comment.

ramonjd committed:
Oh, linter! Thou are more immovable than a wine stain from a woollen mat!

ramonjd commented:
Looks like prod is now `v10.7.4` so we should be good to retest this one. 

```
Gutenberg: v10.7.4
CoBlocks: 2.10.1
Editing Toolkit: dev
WP AMP: 2.0.5
```

ramonjd committed:
Reregistering all core/gutenberg patterns using site locale.
Refactored block patterns class to determine its own source sites.

ramonjd committed:
Adding test to make sure we're making the extra call for the site editor patterns

ramonjd committed:
Adding TODO and other comments

ramonjd committed:
Oh, great Linter, please accept these humble formatting commits as an offering and may you have mercy on us, who prefer `[]` and distrust YODA conditions.

andrewserong reviewed (approved):
Nice work @ramonjd, I like the subtle improvements you've made to this too with adding type hinting, switching out `$patterns_sources` for an `$editor_type` string, and the good inline comments in `reregister_core_and_gutenberg_patterns`.

In practice, since it looks like Gutenberg is already loading the remote patterns in the post and site editors, the main difference after applying this PR is that the Query patterns are now visible in the inserter?

Here's how they appear in the post editor on my test site, where the most recent post in my site does not have a post excerpt or a featured image:

![image](https://user-images.githubusercontent.com/14988353/122719346-67f3b000-d2b1-11eb-96b9-611579ec1270.png)

Unfortunately in this case, the patterns aren't rendered in a very pleasing way. I'm not sure if there is much we can do about this for patterns using the Query block, but just thought I'd mention it because as a new user if I was browsing this category, I'd feel a bit confused about what I'm looking at. Maybe an enhancement for the Featured Image block and Post Excerpt blocks — to make the placeholder state look more approximate to what it would look like if there was valid content? Or should we limit the Query patterns to just show in the site editor?(though I can see them being really useful when editing pages, so it's probably better to have them available everywhere?)

I like that we're surfacing these patterns, though, so don't let these comments be a blocker if we'd like to make them available to users!

ramonjd commented:
> since it looks like Gutenberg is already loading the remote patterns in the post and site editors, the main difference after applying this PR is that the Query patterns are now visible in the inserter?

Yes! And also the other [Gutenberg blocks](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-patterns.php#L149) such as `social-links-shared-background-color` (and others that might appear in the future)

But now I'm thinking whether we should go "full Gutenberg" and not call `_register_core_block_patterns_and_categories` since Gutenberg deregisters these and calls the remote patterns anyway. 

The above is an esoteric point only and doesn't affect the functionality of this PR.

> Unfortunately in this case, the patterns aren't rendered in a very pleasing way. I'm not sure if there is much we can do about this for patterns using the Query block, but just thought I'd mention it because as a new user if I was browsing this category, I'd feel a bit confused about what I'm looking at. 

Agree. It would be nice to have a placeholder for the feature image. 

I think these changes are fairly innocuous. We'll have to revisit the way we register patterns anyway after 5.8, which will load its own patterns remotely.

andrewserong commented:
> But now I'm thinking whether we should go "full Gutenberg" and not call _register_core_block_patterns_and_categories since Gutenberg deregisters these and calls the remote patterns anyway.

That's an interesting point, sounds like it'd be safe to go either way, but if the goal is to primarily focus on the remote patterns, it sounds safe to remove if you wanted to?

> I think these changes are fairly innocuous. We'll have to revisit the way we register patterns anyway after 5.8, which will load its own patterns remotely.

Agreed! The changes do seem fairly innocuous, and it includes a couple of improvements to the code, too 👍

ramonjd committed:
Removing loading core patterns and basically duplicating Gutenberg's pattern loading/registration for now.
The duplication is necessary because we want the patterns to follow the site locale, not the account locale.

ramonjd commented:
Gutenberg loads remote core patterns via https://api.wordpress.org/patterns/1.0/?pattern-keywords=11

If you look at the results of this request in a browser, it's a mixture of languages with a whole bunch of duplicates. Take a look at the headers category:


<img width="323" alt="Screen Shot 2021-07-13 at 3 14 51 pm" src="https://user-images.githubusercontent.com/6458278/125394488-15f40500-e3ed-11eb-9886-2f642c8790e3.png">


It's hard to prevent Gutenberg loading remote core patterns on WPCOM because it's contingent on `has_theme_support( 'core-block-patterns' );`, which core sets as true and Gutenberg access immediately after.

The most reliable way is still to loop through registered patterns and unregister `core/*` patterns.

The good news is that work is under way in Gutenberg to filter patterns by the user locale in https://github.com/WordPress/gutenberg/pull/33052

After chatting with @andrewserong we thought it might be a good idea to rethink our use of the site locale to display pattern translations, giving preference to the user locale, as WordPress Core and Gutenberg (will) do.

We see the following benefits:

- The pattern translations will be in the same language as the UI, which would be the user's preferred language. Categories and descriptions would hence be listed in the user's preferred language. While we can't be sure, we aren't confident that users would use the default text for most of the patterns anyway, or find it important/beneficial to have them translated into the current site language. 
- We can, in the future, use Gutenberg/Core patterns out of the box, with none of the unregister/register fluffery we're trying to do here.

I'm going to close this PR (it's a bit stale anyway) and open up one that will:

- Use the user's locale when fetching patterns
- allow Gutenberg to register its own patterns (which includes Query patterns)
- deregister core remote patterns (until they're filtered)
- reregister WordPress (non-remote) core patterns
